File: 207-CourseSchedule/207-CourseSchedule.js
// Last updated: 4/29/2026, 10:32:59 AM
/**
 * @param {number} numCourses
 * @param {number[][]} prerequisites
 * @return {boolean}
 */
var canFinish = function(numCourses, prerequisites) {
    const adj = {};
    for(const [u, v] of prerequisites) {
        if(adj[v] === undefined) {
            adj[v] = new Set();
        }
        adj[v].add(u);
    }
    
    const checked = new Array(numCourses).fill(false);
    const path = new Array(numCourses).fill(false);
    
    const isCyclic = i => {
        if(checked[i]) {
            return false;
        }
        if(path[i]) {
            return true;
        }
        path[i] = true;
        let ret = false;
        const ends = adj[i];
        if(ends) {
            for(const end of adj[i]) {
                ret = isCyclic(end)
                if(ret) {
                    break;
                }
            }
        }
        path[i] = false;
        checked[i] = true;
        return ret;
    }
    
    for(let i = 0; i < numCourses; i++) {
        if(isCyclic(i)) {
            return false;
        }
    }
    return true;
};

File: 207-CourseSchedule/207-CourseSchedule.py
# Last updated: 4/29/2026, 10:33:03 AM
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        from collections import defaultdict
        adj = defaultdict(set)
        for end, start in prerequisites:
            adj[start].add(end)
        print(adj)
        
        checked = [False] * numCourses
        path = [False] * numCourses
        
        for i in range(numCourses):
            if self.isCyclic(i, adj, checked, path):
                return False
        return True
    
    def isCyclic(self, i, adj, checked, path):
        if checked[i]:
            return False
        if path[i]:
            return True
        path[i] = True
        ret = False
        for child in adj[i]:
            ret = self.isCyclic(child, adj, checked, path)
            if ret:
                break
        
        path[i] = False
        
        checked[i] = True
        return ret
        